runtime-transaction/src/program_id_flags.rs

```diff
@@ -0,0 +1,75 @@
+use {
+    solana_pubkey::Pubkey,
+    solana_sdk::{compute_budget, ed25519_program, secp256k1_program},
+};
+
+/// Top-level program id flags.
+/// The main checking point is done with this enum, and for individual
+/// processors, the flags are converted to specific sub-enums.
+#[derive(Clone)]
+pub enum ProgramIdFlag {
+    NoMatch,
+    Secp256k1,
+    Ed25519,
+    ComputeBudgetProgram,
+}
+
+/// Translate from a [Pubkey] to a [ProgramIdFlag].
+impl From<&Pubkey> for ProgramIdFlag {
+    #[inline]
+    fn from(program_id: &Pubkey) -> Self {
+        if program_id == &secp256k1_program::ID {
+            Self::Secp256k1
+        } else if program_id == &ed25519_program::ID {
+            Self::Ed25519
+        } else if program_id == &compute_budget::ID {
+            Self::ComputeBudgetProgram
+        } else {
+            Self::NoMatch
+        }
+    }
+}
+
+/// Self translation for [ProgramIdFlag].
+impl From<&ProgramIdFlag> for ProgramIdFlag {
+    #[inline]
+    fn from(flag: &ProgramIdFlag) -> Self {
+        flag.clone()
+    }
+}
+
+/// Program flag for determining how to process instructions for finding the
+/// signature details.
+pub enum SignatureDetailsFlag {
+    NoMatch,
```

Comment:
Having to specify `NoMatch` at both levels is a redundancy not exists w Visitor

Reply:
I plan to write some benchmarks in this, but visitor is inefficient because we need to re-check the key for each visitor.

If we add more stuff here, like builtins (which is really my goal rn fro cost-model stuff), then we end up doing more work w/ the visitor pattern.
Let's consider when we add BUILTINs.
I already know from my `compute_budget` and `precompile_details` if an instruction is one of those or not.
If it is one of them, then I don't need to do a hash-table lookup - I already know the cost, because I've already identified the key.

Same even as is. If I see that something is compute-budget key, I don't need to check if it's secp or ed25519, it's not because it's compute-budget.